pkg/iac/rego/store.go

```
@@ -3,26 +3,42 @@ package rego
 import (
 	"fmt"
 	"io/fs"
-	"os"
-	"path/filepath"
+	"path"
 	"slices"
 	"strings"
 
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/aquasecurity/trivy/pkg/set"
 	"github.com/open-policy-agent/opa/v1/loader"
 	"github.com/open-policy-agent/opa/v1/storage"
 )
 
 // initialize a store populated with OPA data files found in dataPaths
 func initStore(dataFS fs.FS, dataPaths, namespaces []string) (storage.Store, error) {
-	// FilteredPaths will recursively find all file paths that contain a valid document
-	// extension from the given list of data paths.
-	allDocumentPaths, _ := loader.FilteredPathsFS(dataFS, dataPaths,
-		func(abspath string, info os.FileInfo, depth int) bool {
-			return !isDataFile(info)
-		},
-	)
-
-	documents, err := loader.NewFileLoader().WithFS(dataFS).All(allDocumentPaths)
+	dataFiles := set.New[string]()
+
+	// The virtual file system uses a slash ('/') as a path separator,
+	// but OPA uses the filepath package, which is OS-dependent.
+	// Therefore, we need to collect all the paths ourselves and pass them to OPA.
+	for _, root := range dataPaths {
+		if err := fs.WalkDir(dataFS, root, func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+			if d.IsDir() {
+				return nil
+			}
+
+			if isDataFile(path) {
+				dataFiles.Append(path)
+			}
+			return nil
+		}); err != nil {
+			log.Error("Failed to collect data file paths", log.String("root", root), log.Err(err))
```

@simar7 I wanted to return an error here but I was failing [this](https://github.com/aquasecurity/trivy/blob/main/pkg/iac/rego/scanner_test.go) test. It was succeeding before because it ignored the error. Was this done premeditated? I find this test strange and it's not clear what it covers. Let me know if it's better to return an error here.

---

Which test are you referring to? The link just points to `scanner_test.go` file.

---

I copied it wrong https://github.com/aquasecurity/trivy/blob/main/pkg/iac/rego/scanner_test.go#L729-L766

---

The current behavior ignores the error (if any) that gets returned by `fs.WalkDir`. The test is simply simulating a case where an invalid filesystem is passed in and therefore `fs.WalkDir` will return an error.

But should we return an error here if we are unable to override the filesystem for data? This would only happen if the passed in `srcFS` isn't nil. In other cases we simply ignore it.

---

`srcFS` is only used to load data if no file system with data is passed through options. Therefore an error can occur in both cases. But I think I understood that the test is just checking that the data was not loaded.